In SearchOperations.searchForStream, 'maxResults' overwrites 'size' (pageSize)

Hi, I'm opening this issue to report something likely to be a bug. 
When creating a `SearchRequest` for `searchForStream` method, `maxResults` parameter is overwriting `size`, while it seems like it shouldn't be.

## Bug
I set both `maxResult` and `Pageable.size` (will refer to `pageSize`) fields for the query and passes it to SearchOperations.searchForStream. 
When `requestConverter.searchRequest`inside the method creates a `SearchRequest`, `maxResult` overwrites `pageSize`.

In the below code, `builder.size` is called first with `query.getPageable().getPageSize()`, followed by another call with `query.getMaxResults()`, overwriting the result of the first call.
https://github.com/spring-projects/spring-data-elasticsearch/blob/0e5af905818ca852c377c0aa576e56516b713dac/src/main/java/org/springframework/data/elasticsearch/client/elc/RequestConverter.java#L1469-L1492

I was stuck at the case where,
- 250,000 data available in the ES index
- Want to fetch at most 150,000 documents in total (out of 250k), using `SearchOperations.searchForStream` (`maxResult 150000`)
- Want to fetch 10,000 documents per page (per scroll search request), which fills almost all the response memory buffer (`Pageable.ofSize(10000)`)

result
- Expected : 15 search requests in total, a single search returning a page of 10k documents
- Actual : a single search tries to fetch 150k documents, overflowing response memory buffer and causing below Exception

`Caused by: org.springframework.dao.DataAccessResourceFailureException: entity content is too long [169386197] for the configured buffer limit [104857600]; nested exception is java.lang.RuntimeException: entity content is too long [169386197] for the configured buffer limit [104857600]`

<br>

## Possible solution
I think two parameters have totally separate roles, thus one shouldn't overwrite the other.
- `pageSize` parameter should decide how many documents that I want ES to fetch per page, 
- while `maxResult` parameter decides how many documents I want to fetch **_in total_** during the whole stream operation. 

👉 Therefore, the actual pageSize should be `Min(maxResult, pageSize)` if `maxResult` is set

Changes made : https://github.com/hy2850/spring-data-elasticsearch/commits/%233089-size-and-maxResults/

<br>

## Further explanation with code for your understanding
`SearchOperations.searchForStream` returns a CloseableIterator that uses scroll API to search through a large set of data by pages.
Currently, `maxResult` is solely used to get `maxCount` for this method.

https://github.com/spring-projects/spring-data-elasticsearch/blob/0e5af905818ca852c377c0aa576e56516b713dac/src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java#L410-L423

The iterator keeps track of how many documents it has fetched so far, in `currentCount` instance variable.
https://github.com/spring-projects/spring-data-elasticsearch/blob/0e5af905818ca852c377c0aa576e56516b713dac/src/main/java/org/springframework/data/elasticsearch/core/StreamQueries.java#L130-L137

When `currentCount` reaches`maxCount`, the iterator is forced to return false for `hasNext`, even when there could be more documents available for searching. 
https://github.com/spring-projects/spring-data-elasticsearch/blob/0e5af905818ca852c377c0aa576e56516b713dac/src/main/java/org/springframework/data/elasticsearch/core/StreamQueries.java#L107-L116

So, I reached a conclusion that `maxResult` should only decide `maxCount`, not the `pageSize`, and this is a bug.

--- a/src/main/java/org/springframework/data/elasticsearch/client/elc/RequestConverter.java
+++ b/src/main/java/org/springframework/data/elasticsearch/client/elc/RequestConverter.java
@@ -1295,11 +1295,15 @@ class RequestConverter extends AbstractQueryProcessor {
 									.timeout(timeStringMs(query.getTimeout())) //
 							;
 
-							if (query.getPageable().isPaged()) {
-								bb //
-										.from((int) query.getPageable().getOffset()) //
-										.size(query.getPageable().getPageSize());
-							}
+							var offset = query.getPageable().isPaged() ? query.getPageable().getOffset() : 0;
+							var pageSize = query.getPageable().isPaged() ? query.getPageable().getPageSize()
+									: INDEX_MAX_RESULT_WINDOW;
+							// if we have both a page size and a max results, we take the min, this is necessary for
+							// searchForStream to work correctly (#3098) as there the page size defines what is
+							// returned in a single request, and the max result determines the total number of
+							// documents returned
+							var size = query.isLimiting() ? Math.min(pageSize, query.getMaxResults()) : pageSize;
+							bb.from((int) offset).size(size);
 
 							if (!isEmpty(query.getFields())) {
 								bb.fields(fb -> {
@@ -1310,10 +1314,6 @@ class RequestConverter extends AbstractQueryProcessor {
 
 							if (!isEmpty(query.getStoredFields())) {
 								bb.storedFields(query.getStoredFields());
-							}
-
-							if (query.isLimiting()) {
-								bb.size(query.getMaxResults());
 							}
 
 							if (query.getMinScore() > 0) {
@@ -1473,13 +1473,14 @@ class RequestConverter extends AbstractQueryProcessor {
 			builder.seqNoPrimaryTerm(true);
 		}
 
-		if (query.getPageable().isPaged()) {
-			builder //
-					.from((int) query.getPageable().getOffset()) //
-					.size(query.getPageable().getPageSize());
-		} else {
-			builder.from(0).size(INDEX_MAX_RESULT_WINDOW);
-		}
+		var offset = query.getPageable().isPaged() ? query.getPageable().getOffset() : 0;
+		var pageSize = query.getPageable().isPaged() ? query.getPageable().getPageSize() : INDEX_MAX_RESULT_WINDOW;
+		// if we have both a page size and a max results, we take the min, this is necessary for
+		// searchForStream to work correctly (#3098) as there the page size defines what is
+		// returned in a single request, and the max result determines the total number of
+		// documents returned
+		var size = query.isLimiting() ? Math.min(pageSize, query.getMaxResults()) : pageSize;
+		builder.from((int) offset).size(size);
 
 		if (!isEmpty(query.getFields())) {
 			var fieldAndFormats = query.getFields().stream().map(field -> FieldAndFormat.of(b -> b.field(field))).toList();
@@ -1492,10 +1493,6 @@ class RequestConverter extends AbstractQueryProcessor {
 
 		if (query.getIndicesOptions() != null) {
 			addIndicesOptions(builder, query.getIndicesOptions());
-		}
-
-		if (query.isLimiting()) {
-			builder.size(query.getMaxResults());
 		}
 
 		if (query.getMinScore() > 0) {

